### ISSUE TITLE
 Is it possible to specify a subset of nodes in a graph from which the loss function will be calculated?

### ISSUE THREAD

**Issue by ognjenkundacina** [BNN-UPC/ignnition#83]
Dear ignnition team,

In our problem we use a GNN to learn over graphs with two type of nodes. One type of nodes is intended to be used for inputting data into the GNN, and the other type is used for generating predictions (no nodes are intended for both predictions and inputting data). Furthermore, these graphs are bipartite. So in the ideal case we would like to label only the second type of nodes and calculate the loss functions using their labels and predictions.

Is it possible to specify a subset of nodes in a graph (type of nodes in our case) from which the loss function will be calculated?

Thanks!

Ognjen

**Comment by jsuarezv** [comment]
Dear Ognjen,
Thank you for your interest. What you mention is possible with IGNNITION. You just need to define two different entities (i.e., element types), and then define only outputs for one of them. If you want to add input features only for one element type, you can initialize the hidden states of the other entity without any feature (i.e.,  "input: [ ]"), so the corresponding hidden state vectors are automatically zero-padded at the beginning.

I think the following example can be useful for you:
https://github.com/BNN-UPC/ignnition/blob/main/examples/Routenet/model_description.yaml

RouteNet operates over a bipartite graph, where graph elements can be paths or links, and it finally outputs only values on paths, so the loss is only computed on these final path-level predictions.
 

Regards,
José

**Comment by MiquelFerriol** [comment]
Hi @ognjenkundacina,
Yes! It is possible.

Let's say you have two types of identities: once for "inputting" the data called input, and another one called "output" to generate the predictions. Let's also suppose that the "input" has as parameter a variable called "x" and the value to be predicted is called "y". Then, the entity description found in the model_description.yaml should look something like this:
```yaml
- name: input
  state_dimension: 32
  initial_state:
    - type: build_state
      input: [$x]

- name: output
  state_dimension: 32
```

 The specification over which nodes the GNN needs to make the predictions is defined in the readout phase. In this particular case (let's suppose the readout function is a neural network), you could specify it as:
```yaml
readout:
- type: neural_network
  input: [output]
  nn_name: readout_model
  output_label: [$y]
```

Note that the way of indicating over which nodes the prediction is done and the loss is computed is by defining the output_label. In this particular case, the model will only predict the "y" value of the "output" entity.

[Here](https://github.com/BNN-UPC/ignnition/tree/main/examples/Routenet) you can find a similar example of what you are trying to do. In it, the graph considers two types of nodes: links and paths. However, the predictions are only done over the paths.

Hope it helps!
Regards,
Miquel

**Comment by ognjenkundacina** [comment]
Dear José and Miquel,

Thank you for your detailed answers!

Regards,
Ognjen
